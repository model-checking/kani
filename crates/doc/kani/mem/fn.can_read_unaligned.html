<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Checks that pointer `ptr` point to a valid value of type `T`."><title>can_read_unaligned in kani::mem - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="kani" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (1d23d0680 2025-10-22)" data-channel="nightly" data-search-js="search-d69d8955.js" data-stringdex-js="stringdex-c3e638e9.js" data-settings-js="settings-c38705f0.js" ><script src="../../static.files/storage-e2aeef58.js"></script><script defer src="sidebar-items.js"></script><script defer src="../../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc fn"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">can_read_unaligned</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../kani/index.html">kani</a><span class="version">0.65.0</span></h2></div><div class="sidebar-elems"><div id="rustdoc-modnav"><h2><a href="index.html">In kani::<wbr>mem</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">kani</a>::<wbr><a href="index.html">mem</a></div><h1>Function <span class="fn">can_<wbr>read_<wbr>unaligned</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/kani/lib.rs.html#57">Source</a> </span></div><pre class="rust item-decl"><code>pub fn can_read_unaligned&lt;T: <a class="trait" href="https://doc.rust-lang.org/nightly/core/marker/trait.MetaSized.html" title="trait core::marker::MetaSized">MetaSized</a>&gt;(ptr: <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.pointer.html">*const T</a>) -&gt; <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.bool.html">bool</a></code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Checks that pointer <code>ptr</code> point to a valid value of type <code>T</code>.</p>
<p>For that, the pointer has to be a valid pointer according to <a href="index.html" title="mod kani::mem">crate::mem</a> conditions 1, 2
and 3,
and the value stored must respect the validity invariants for type <code>T</code>.</p>
<p>Note this function succeeds for unaligned pointers. See <a href="fn.can_dereference.html" title="fn kani::mem::can_dereference">self::can_dereference</a> if you also
want to check pointer alignment.</p>
<p>This function will panic today if the pointer is not null, and it points to an unallocated or
deallocated memory location. This is an existing Kani limitation.
See <a href="https://github.com/model-checking/kani/issues/2690">https://github.com/model-checking/kani/issues/2690</a> for more details.</p>
</div></details></section></div></main></body></html>