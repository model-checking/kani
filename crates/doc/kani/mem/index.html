<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="This module contains functions useful for checking unsafe memory access."><title>kani::mem - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="kani" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (1d23d0680 2025-10-22)" data-channel="nightly" data-search-js="search-d69d8955.js" data-stringdex-js="stringdex-c3e638e9.js" data-settings-js="settings-c38705f0.js" ><script src="../../static.files/storage-e2aeef58.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Module mem</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../kani/index.html">kani</a><span class="version">0.65.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module mem</a></h2><h3><a href="#functions">Module Items</a></h3><ul class="block"><li><a href="#functions" title="Functions">Functions</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate kani</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">kani</a></div><h1>Module <span>mem</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/kani/lib.rs.html#57">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>This module contains functions useful for checking unsafe memory access.</p>
<p>Given the following validity rules provided in the Rust documentation:
<a href="https://doc.rust-lang.org/std/ptr/index.html">https://doc.rust-lang.org/std/ptr/index.html</a> (accessed May 20th, 2025)</p>
<ol>
<li>For memory accesses of size zero, every pointer is valid, including the null pointer.
The following points are only concerned with non-zero-sized accesses.</li>
<li>A null pointer is never valid.</li>
<li>For a pointer to be valid, it is necessary, but not always sufficient, that the pointer
be dereferenceable: the memory range of the given size starting at the pointer must all be
within the bounds of a single allocated object. Note that in Rust, every (stack-allocated)
variable is considered a separate allocated object.</li>
<li>All accesses performed by functions in this module are non-atomic in the sense of atomic
operations used to synchronize between threads.
This means it is undefined behavior to perform two concurrent accesses to the same location
from different threads unless both accesses only read from memory.
Notice that this explicitly includes <code>read_volatile</code> and <code>write_volatile</code>:
Volatile accesses cannot be used for inter-thread synchronization.</li>
<li>The result of casting a reference to a pointer is valid for as long as the underlying
object is live and no reference (just raw pointers) is used to access the same memory.
That is, reference and pointer accesses cannot be interleaved.</li>
</ol>
<p>Kani is able to verify #1, #2, and #3 today.</p>
</div></details><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="fn" href="fn.can_dereference.html" title="fn kani::mem::can_dereference">can_<wbr>dereference</a></dt><dd>Checks that pointer <code>ptr</code> point to a valid value of type <code>T</code>.</dd><dt><a class="fn" href="fn.can_read_unaligned.html" title="fn kani::mem::can_read_unaligned">can_<wbr>read_<wbr>unaligned</a></dt><dd>Checks that pointer <code>ptr</code> point to a valid value of type <code>T</code>.</dd><dt><a class="fn" href="fn.can_write.html" title="fn kani::mem::can_write">can_<wbr>write</a></dt><dd>Check if the pointer is valid for write access according to <a href="index.html" title="mod kani::mem">crate::mem</a> conditions 1, 2
and 3.</dd><dt><a class="fn" href="fn.can_write_unaligned.html" title="fn kani::mem::can_write_unaligned">can_<wbr>write_<wbr>unaligned</a></dt><dd>Check if the pointer is valid for unaligned write access according to <a href="index.html" title="mod kani::mem">crate::mem</a> conditions
1, 2 and 3.</dd><dt><a class="fn" href="fn.checked_align_of_raw.html" title="fn kani::mem::checked_align_of_raw">checked_<wbr>align_<wbr>of_<wbr>raw</a></dt><dd>Compute the size of the val pointed to if safe.</dd><dt><a class="fn" href="fn.checked_size_of_raw.html" title="fn kani::mem::checked_size_of_raw">checked_<wbr>size_<wbr>of_<wbr>raw</a></dt><dd>Compute the size of the val pointed to if it is safe to do so.</dd><dt><a class="fn" href="fn.is_inbounds.html" title="fn kani::mem::is_inbounds">is_<wbr>inbounds</a></dt><dd>Checks that <code>ptr</code> points to an allocation that can hold data of size calculated from <code>T</code>.</dd><dt><a class="fn" href="fn.same_allocation.html" title="fn kani::mem::same_allocation">same_<wbr>allocation</a></dt><dd>Check if two pointers points to the same allocated object, and that both pointers
are in bounds of that object.</dd></dl></section></div></main></body></html>