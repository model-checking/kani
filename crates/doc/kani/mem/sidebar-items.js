window.SIDEBAR_ITEMS = {"fn":["can_dereference","can_read_unaligned","can_write","can_write_unaligned","checked_align_of_raw","checked_size_of_raw","is_inbounds","same_allocation"]};