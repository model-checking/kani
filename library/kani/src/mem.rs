// Copyright Kani Contributors
// SPDX-License-Identifier: Apache-2.0 OR MIT
//! This module contains functions useful for checking unsafe memory access.
//!
//! Given the following validity rules provided in the Rust documentation:
//! <https://doc.rust-lang.org/std/ptr/index.html> (accessed Feb 6th, 2024)
//!
//! 1. A null pointer is never valid, not even for accesses of size zero.
//! 2. For a pointer to be valid, it is necessary, but not always sufficient, that the pointer
//!    be dereferenceable: the memory range of the given size starting at the pointer must all be
//!    within the bounds of a single allocated object. Note that in Rust, every (stack-allocated)
//!    variable is considered a separate allocated object.
//!    ~~Even for operations of size zero, the pointer must not be pointing to deallocated memory,
//!    i.e., deallocation makes pointers invalid even for zero-sized operations.~~
//!    ZST access is not OK for any pointer.
//!    See: <https://github.com/rust-lang/unsafe-code-guidelines/issues/472>
//! 3. However, casting any non-zero integer literal to a pointer is valid for zero-sized
//!    accesses, even if some memory happens to exist at that address and gets deallocated.
//!    This corresponds to writing your own allocator: allocating zero-sized objects is not very
//!    hard. The canonical way to obtain a pointer that is valid for zero-sized accesses is
//!    `NonNull::dangling`.
//! 4. All accesses performed by functions in this module are non-atomic in the sense of atomic
//!    operations used to synchronize between threads.
//!    This means it is undefined behavior to perform two concurrent accesses to the same location
//!    from different threads unless both accesses only read from memory.
//!    Notice that this explicitly includes `read_volatile` and `write_volatile`:
//!    Volatile accesses cannot be used for inter-thread synchronization.
//! 5. The result of casting a reference to a pointer is valid for as long as the underlying
//!    object is live and no reference (just raw pointers) is used to access the same memory.
//!    That is, reference and pointer accesses cannot be interleaved.
//!
//! Kani is able to verify #1 and #2 today.
//!
//! For #3, we are overly cautious, and Kani will only consider zero-sized pointer access safe if
//! the address matches `NonNull::<()>::dangling()`.
//! The way Kani tracks provenance is not enough to check if the address was the result of a cast
//! from a non-zero integer literal.

use crate::kani_intrinsic;
use crate::mem::private::Internal;
use std::mem::{align_of, size_of};
use std::ptr::{DynMetadata, NonNull, Pointee};

/// Check if the pointer is valid for write access according to [crate::mem] conditions 1, 2
/// and 3.
///
/// Note this function also checks for pointer alignment. Use [self::can_write_unaligned]
/// if you don't want to fail for unaligned pointers.
///
/// This function does not check if the value stored is valid for the given type. Use
/// [self::can_dereference] for that.
///
/// This function will panic today if the pointer is not null, and it points to an unallocated or
/// deallocated memory location. This is an existing Kani limitation.
/// See <https://github.com/model-checking/kani/issues/2690> for more details.
#[crate::unstable(
    feature = "mem-predicates",
    issue = 2690,
    reason = "experimental memory predicate API"
)]
pub fn can_write<T>(ptr: *mut T) -> bool
where
    T: ?Sized,
    <T as Pointee>::Metadata: PtrProperties<T>,
{
    // The interface takes a mutable pointer to improve readability of the signature.
    // However, using constant pointer avoid unnecessary instrumentation, and it is as powerful.
    // Hence, cast to `*const T`.
    let ptr: *const T = ptr;
    let (thin_ptr, metadata) = ptr.to_raw_parts();
    metadata.is_ptr_aligned(thin_ptr, Internal) && is_inbounds(&metadata, thin_ptr)
}

/// Check if the pointer is valid for unaligned write access according to [crate::mem] conditions
/// 1, 2 and 3.
///
/// Note this function succeeds for unaligned pointers. See [self::can_write] if you also
/// want to check pointer alignment.
///
/// This function will panic today if the pointer is not null, and it points to an unallocated or
/// deallocated memory location. This is an existing Kani limitation.
/// See <https://github.com/model-checking/kani/issues/2690> for more details.
#[crate::unstable(
    feature = "mem-predicates",
    issue = 2690,
    reason = "experimental memory predicate API"
)]
pub fn can_write_unaligned<T>(ptr: *const T) -> bool
where
    T: ?Sized,
    <T as Pointee>::Metadata: PtrProperties<T>,
{
    let (thin_ptr, metadata) = ptr.to_raw_parts();
    is_inbounds(&metadata, thin_ptr)
}

/// Checks that pointer `ptr` point to a valid value of type `T`.
///
/// For that, the pointer has to be a valid pointer according to [crate::mem] conditions 1, 2
/// and 3,
/// and the value stored must respect the validity invariants for type `T`.
///
/// TODO: Kani should automatically add those checks when a de-reference happens.
/// <https://github.com/model-checking/kani/issues/2975>
///
/// This function will panic today if the pointer is not null, and it points to an unallocated or
/// deallocated memory location. This is an existing Kani limitation.
/// See <https://github.com/model-checking/kani/issues/2690> for more details.
#[crate::unstable(
    feature = "mem-predicates",
    issue = 2690,
    reason = "experimental memory predicate API"
)]
#[allow(clippy::not_unsafe_ptr_arg_deref)]
pub fn can_dereference<T>(ptr: *const T) -> bool
where
    T: ?Sized,
    <T as Pointee>::Metadata: PtrProperties<T>,
{
    let (thin_ptr, metadata) = ptr.to_raw_parts();
    // Need to assert `is_initialized` because non-determinism is used under the hood, so it does
    // not make sense to use it inside assumption context.
    metadata.is_ptr_aligned(thin_ptr, Internal)
        && is_inbounds(&metadata, thin_ptr)
        && assert_is_initialized(ptr)
        && unsafe { has_valid_value(ptr) }
}

/// Checks that pointer `ptr` point to a valid value of type `T`.
///
/// For that, the pointer has to be a valid pointer according to [crate::mem] conditions 1, 2
/// and 3,
/// and the value stored must respect the validity invariants for type `T`.
///
/// Note this function succeeds for unaligned pointers. See [self::can_dereference] if you also
/// want to check pointer alignment.
///
/// This function will panic today if the pointer is not null, and it points to an unallocated or
/// deallocated memory location. This is an existing Kani limitation.
/// See <https://github.com/model-checking/kani/issues/2690> for more details.
#[crate::unstable(
    feature = "mem-predicates",
    issue = 2690,
    reason = "experimental memory predicate API"
)]
#[allow(clippy::not_unsafe_ptr_arg_deref)]
pub fn can_read_unaligned<T>(ptr: *const T) -> bool
where
    T: ?Sized,
    <T as Pointee>::Metadata: PtrProperties<T>,
{
    let (thin_ptr, metadata) = ptr.to_raw_parts();
    // Need to assert `is_initialized` because non-determinism is used under the hood, so it does
    // not make sense to use it inside assumption context.
    is_inbounds(&metadata, thin_ptr)
        && assert_is_initialized(ptr)
        && unsafe { has_valid_value(ptr) }
}

/// Checks that `data_ptr` points to an allocation that can hold data of size calculated from `T`.
///
/// This will panic if `data_ptr` points to an invalid `non_null`
fn is_inbounds<M, T>(metadata: &M, data_ptr: *const ()) -> bool
where
    M: PtrProperties<T>,
    T: ?Sized,
{
    let sz = metadata.pointee_size(Internal);
    if sz == 0 {
        true // ZST pointers are always valid including nullptr.
    } else if data_ptr.is_null() {
        false
    } else {
        // Note that this branch can't be tested in concrete execution as `is_read_ok` needs to be
        // stubbed.
        // We first assert that the data_ptr
        crate::assert(
            unsafe { is_allocated(data_ptr, 0) },
            "Kani does not support reasoning about pointer to unallocated memory",
        );
        unsafe { is_allocated(data_ptr, sz) }
    }
}

mod private {
    /// Define like this to restrict usage of PtrProperties functions outside Kani.
    #[derive(Copy, Clone)]
    pub struct Internal;
}

/// Trait that allow us to extract information from pointers without de-referencing them.
#[doc(hidden)]
pub trait PtrProperties<T: ?Sized> {
    fn pointee_size(&self, _: Internal) -> usize;

    /// A pointer is aligned if its address is a multiple of its minimum alignment.
    fn is_ptr_aligned(&self, ptr: *const (), internal: Internal) -> bool {
        let min = self.min_alignment(internal);
        ptr as usize % min == 0
    }

    fn min_alignment(&self, _: Internal) -> usize;

    fn dangling(&self, _: Internal) -> *const ();
}

/// Get the information for sized types (they don't have metadata).
impl<T> PtrProperties<T> for () {
    fn pointee_size(&self, _: Internal) -> usize {
        size_of::<T>()
    }

    fn min_alignment(&self, _: Internal) -> usize {
        align_of::<T>()
    }

    fn dangling(&self, _: Internal) -> *const () {
        NonNull::<T>::dangling().as_ptr() as *const _
    }
}

/// Get the information from the str metadata.
impl PtrProperties<str> for usize {
    #[inline(always)]
    fn pointee_size(&self, _: Internal) -> usize {
        *self
    }

    /// String slices are a UTF-8 representation of characters that have the same layout as slices
    /// of type [u8].
    /// <https://doc.rust-lang.org/reference/type-layout.html#str-layout>
    fn min_alignment(&self, _: Internal) -> usize {
        align_of::<u8>()
    }

    fn dangling(&self, _: Internal) -> *const () {
        NonNull::<u8>::dangling().as_ptr() as _
    }
}

/// Get the information from the slice metadata.
impl<T> PtrProperties<[T]> for usize {
    fn pointee_size(&self, _: Internal) -> usize {
        *self * size_of::<T>()
    }

    fn min_alignment(&self, _: Internal) -> usize {
        align_of::<T>()
    }

    fn dangling(&self, _: Internal) -> *const () {
        NonNull::<T>::dangling().as_ptr() as _
    }
}

/// Get the information from the vtable.
impl<T> PtrProperties<T> for DynMetadata<T>
where
    T: ?Sized,
{
    fn pointee_size(&self, _: Internal) -> usize {
        self.size_of()
    }

    fn min_alignment(&self, _: Internal) -> usize {
        self.align_of()
    }

    fn dangling(&self, _: Internal) -> *const () {
        NonNull::<&T>::dangling().as_ptr() as _
    }
}

/// Check if the pointer `_ptr` contains an allocated address of size equal or greater than `_size`.
///
/// # Safety
///
/// This function should only be called to ensure a pointer is always valid, i.e., in an assertion
/// context.
///
/// I.e.: This function always returns `true` if the pointer is valid.
/// Otherwise, it returns non-det boolean.
#[rustc_diagnostic_item = "KaniIsAllocated"]
#[inline(never)]
unsafe fn is_allocated(_ptr: *const (), _size: usize) -> bool {
    kani_intrinsic()
}

/// Check if the value stored in the given location satisfies type `T` validity requirements.
///
/// # Safety
///
/// - Users have to ensure that the pointer is aligned the pointed memory is allocated.
#[rustc_diagnostic_item = "KaniValidValue"]
#[inline(never)]
unsafe fn has_valid_value<T: ?Sized>(_ptr: *const T) -> bool {
    kani_intrinsic()
}

/// Check whether `len * size_of::<T>()` bytes are initialized starting from `ptr`.
#[rustc_diagnostic_item = "KaniIsInitialized"]
#[inline(never)]
pub(crate) fn is_initialized<T: ?Sized>(_ptr: *const T) -> bool {
    kani_intrinsic()
}

/// A helper to assert `is_initialized` to use it as a part of other predicates.
fn assert_is_initialized<T: ?Sized>(ptr: *const T) -> bool {
    crate::check(is_initialized(ptr), "Undefined Behavior: Reading from an uninitialized pointer");
    true
}

/// Get the object ID of the given pointer.
#[doc(hidden)]
#[crate::unstable(
    feature = "ghost-state",
    issue = 3184,
    reason = "experimental ghost state/shadow memory API"
)]
#[rustc_diagnostic_item = "KaniPointerObject"]
#[inline(never)]
pub fn pointer_object<T: ?Sized>(_ptr: *const T) -> usize {
    kani_intrinsic()
}

/// Get the object offset of the given pointer.
#[doc(hidden)]
#[crate::unstable(
    feature = "ghost-state",
    issue = 3184,
    reason = "experimental ghost state/shadow memory API"
)]
#[rustc_diagnostic_item = "KaniPointerOffset"]
#[inline(never)]
pub fn pointer_offset<T: ?Sized>(_ptr: *const T) -> usize {
    kani_intrinsic()
}

#[cfg(test)]
mod tests {
    use super::{can_dereference, can_write, PtrProperties};
    use crate::mem::private::Internal;
    use std::fmt::Debug;
    use std::intrinsics::size_of;
    use std::mem::{align_of, align_of_val, size_of_val};
    use std::ptr;
    use std::ptr::{NonNull, Pointee};

    fn size_of_t<T>(ptr: *const T) -> usize
    where
        T: ?Sized,
        <T as Pointee>::Metadata: PtrProperties<T>,
    {
        let (_, metadata) = ptr.to_raw_parts();
        metadata.pointee_size(Internal)
    }

    fn align_of_t<T>(ptr: *const T) -> usize
    where
        T: ?Sized,
        <T as Pointee>::Metadata: PtrProperties<T>,
    {
        let (_, metadata) = ptr.to_raw_parts();
        metadata.min_alignment(Internal)
    }

    #[test]
    fn test_size_of() {
        assert_eq!(size_of_t("hi"), size_of_val("hi"));
        assert_eq!(size_of_t(&0u8), size_of_val(&0u8));
        assert_eq!(size_of_t(&0u8 as *const dyn std::fmt::Display), size_of_val(&0u8));
        assert_eq!(size_of_t(&[0u8, 1u8] as &[u8]), size_of_val(&[0u8, 1u8]));
        assert_eq!(size_of_t(&[] as &[u8]), size_of_val::<[u8; 0]>(&[]));
        assert_eq!(
            size_of_t(NonNull::<u32>::dangling().as_ptr() as *const dyn std::fmt::Display),
            size_of::<u32>()
        );
    }

    #[test]
    fn test_alignment() {
        assert_eq!(align_of_t("hi"), align_of_val("hi"));
        assert_eq!(align_of_t(&0u8), align_of_val(&0u8));
        assert_eq!(align_of_t(&0u32 as *const dyn std::fmt::Display), align_of_val(&0u32));
        assert_eq!(align_of_t(&[0isize, 1isize] as &[isize]), align_of_val(&[0isize, 1isize]));
        assert_eq!(align_of_t(&[] as &[u8]), align_of_val::<[u8; 0]>(&[]));
        assert_eq!(
            align_of_t(NonNull::<u32>::dangling().as_ptr() as *const dyn std::fmt::Display),
            align_of::<u32>()
        );
    }

    #[test]
    pub fn test_empty_slice() {
        let slice_ptr = Vec::<char>::new().as_mut_slice() as *mut [char];
        assert!(can_write(slice_ptr));
    }

    #[test]
    pub fn test_empty_str() {
        let slice_ptr = String::new().as_mut_str() as *mut str;
        assert!(can_write(slice_ptr));
    }

    #[test]
    fn test_dangling_zst() {
        test_dangling_of_zst::<()>();
        test_dangling_of_zst::<[(); 10]>();
    }

    fn test_dangling_of_zst<T>() {
        let dangling: *mut T = NonNull::<T>::dangling().as_ptr();
        assert!(can_write(dangling));

        let vec_ptr = Vec::<T>::new().as_mut_ptr();
        assert!(can_write(vec_ptr));
    }

    #[test]
    fn test_null_fat_ptr() {
        assert!(!can_dereference(ptr::null::<char>() as *const dyn Debug));
    }

    #[test]
    fn test_null_char() {
        assert!(!can_dereference(ptr::null::<char>()));
    }

    #[test]
    fn test_null_mut() {
        assert!(!can_write(ptr::null_mut::<String>()));
    }
}
