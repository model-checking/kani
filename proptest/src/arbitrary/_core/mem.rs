//-
// Copyright 2017, 2018 The proptest developers
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

// SPDX-License-Identifier: Apache-2.0 OR MIT
//
// Modifications Copyright Kani Contributors
// See GitHub history for details.

//! Arbitrary implementations for `std::mem`.

use core::mem::*;

use crate::arbitrary::*;
use crate::strategy::statics::static_map;

arbitrary!([A: Arbitrary] Discriminant<A>,
    SMapped<A, Self>, A::Parameters;
    args => static_map(any_with::<A>(args), |x| discriminant(&x))
);

lift1!(['static] Discriminant<A>;
    base => static_map(base, |x| discriminant(&x))
);

// Not supported at the moment since the user won't be able to call
// https://doc.rust-lang.org/nightly/std/mem/union.ManuallyDrop.html#method.drop
// in any case so the use case is not great for this.
//wrap_ctor!(ManuallyDrop);

#[cfg(all(test, not(kani)))]
mod test {
    #[derive(Copy, Clone, Debug)]
    struct DummyStruct;
    arbitrary!(DummyStruct; DummyStruct);

    no_panic_test!(
        //manually_drop       => ManuallyDrop<u8>, // Trivial destructor.
        discriminant_struct => Discriminant<super::DummyStruct>,
        discriminant_enum   => Discriminant<::std::num::FpCategory>
    );
}
