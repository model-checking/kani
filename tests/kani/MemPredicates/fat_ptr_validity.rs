// Copyright Kani Contributors
// SPDX-License-Identifier: Apache-2.0 OR MIT
// kani-flags: -Z mem-predicates
//! Check that Kani's memory predicates work for fat pointers.

extern crate kani;

use kani::mem::assert_valid_ptr;

mod valid_access {
    use super::*;
    #[kani::proof]
    pub fn check_valid_dyn_ptr() {
        let boxed: Box<dyn PartialEq<u8>> = Box::new(10u8);
        let raw_ptr = Box::into_raw(boxed);
        assert_valid_ptr(raw_ptr);
        let boxed = unsafe { Box::from_raw(raw_ptr) };
        assert_ne!(*boxed, 0);
    }

    #[kani::proof]
    pub fn check_valid_slice_ptr() {
        let array = ['a', 'b', 'c'];
        let slice = &array as *const [char];
        assert_valid_ptr(slice);
        assert_eq!(unsafe { &*slice }[0], 'a');
        assert_eq!(unsafe { &*slice }[2], 'c');
    }

    #[kani::proof]
    pub fn check_valid_zst() {
        let slice_ptr = Vec::<char>::new().as_slice() as *const [char];
        assert_valid_ptr(slice_ptr);

        let str_ptr = String::new().as_str() as *const str;
        assert_valid_ptr(str_ptr);
    }
}

mod invalid_access {
    use super::*;
    #[kani::proof]
    #[kani::should_panic]
    pub fn check_invalid_dyn_ptr() {
        let raw_ptr: *const dyn PartialEq<u8> = unsafe { new_dead_ptr::<u8>(0) };
        assert_valid_ptr(raw_ptr);
    }

    #[kani::proof]
    #[kani::should_panic]
    pub fn check_invalid_slice_ptr() {
        let raw_ptr: *const [char] = unsafe { new_dead_ptr::<[char; 2]>(['a', 'b']) };
        assert_valid_ptr(raw_ptr);
    }

    #[kani::proof]
    #[kani::should_panic]
    pub fn check_invalid_slice_len() {
        let array = [10usize; 10];
        let invalid: *const [usize; 11] = &array as *const [usize; 10] as *const [usize; 11];
        let ptr: *const [usize] = invalid as *const _;
        assert_valid_ptr(ptr);
    }

    unsafe fn new_dead_ptr<T>(val: T) -> *const T {
        let local = val;
        &local as *const _
    }
}
