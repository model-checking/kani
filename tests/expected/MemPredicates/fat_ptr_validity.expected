Failed Checks: Kani does not support reasoning about pointer to unallocated memory

Verification failed for - invalid_access::check_invalid_slice_ptr
Verification failed for - invalid_access::check_invalid_dyn_ptr
Complete - 4 successfully verified harnesses, 2 failures, 6 total.
